### ISSUE TITLE
TC search fails spuriously for primitive projections in 8.20

### ISSUE THREAD

**Issue by Janno** [coq/coq#19668]
### Description of the problem

The example below works in 8.19 but not in 8.20. There are zero candidates so this points to a dnet issue.

`git bisect` points to
a366091573afce31e1cfb5802d2048287f07163d is the first bad commit
```
commit a366091573afce31e1cfb5802d2048287f07163d
Author: Gaëtan Gilbert <gaetan.gilbert@skyskimmer.net>
Date:   Tue Mar 26 15:19:45 2024 +0100

    Define Tacred.soft_evaluable_of_global_reference

 plugins/ltac/tacintern.ml | 20 +-------------------
 pretyping/tacred.ml       | 12 +++++++++++-
 pretyping/tacred.mli      |  8 +++++++-
 vernac/comHints.ml        |  9 ++-------
 4 files changed, 21 insertions(+), 28 deletions(-)
```

cc @SkySkimmer 

### Small Coq file to reproduce the bug

```coq
#[projections(primitive)]
Record bi : Type := BI {car : Type; foo : car}.
Axiom PROP:bi.
Inductive Fwd (P : car PROP) : Type := MKFWD.
Create HintDb db discriminated.
#[local] Hint Opaque foo : db.
#[global] Hint Extern  0 (Fwd ((@foo PROP))) => constructor : db.
Goal Fwd (@foo PROP).
Proof.
  intros.
  Set Printing All.
  Set Typeclasses Debug Verbosity 2.
  typeclasses eauto with db.
Qed.
```

### Version of Coq where this bug occurs

8.20

### Last version of Coq where the bug did not occur

8.19

**Comment by Janno** [comment]
I am not sure I understand. I thought `Hint Opaque proj` had targeted the projection (not the constant) starting with #18327.

**Comment by Janno** [comment]
Your understanding of the commit is indeed correct. What is bothering me to is the fact that we are internally using a patched version of 8.19.1 that includes the fixes around primitive projections and opacity (https://github.com/coq/coq/pull/18785 and https://github.com/coq/coq/pull/18809), emits `EvalProjectionRef` for projections, but does not fail on the example. I am still trying to investigate how that is possible. Our version is in this tag: https://github.com/rlepigre/coq/tree/V8.19.1%2Bbedrock.
